File: 402-remove-k-digits/remove-k-digits.java
class Solution {
    public String removeKdigits(String num, int k) {
        int len = num.length();
        if (k == len) return "0";

        Deque<Character> stack = new ArrayDeque<>();

        for (int i = 0; i < len; i++) {
            char c = num.charAt(i);
            while (!stack.isEmpty() && k > 0 && stack.peekLast() > c) {
                stack.pollLast();
                k--;
            }
            stack.addLast(c);
        }

        while (k-- > 0) {
            stack.pollLast();
        }
        StringBuilder sb = new StringBuilder();
        for (char digit : stack) {
            sb.append(digit);
        }

        while (sb.length() > 0 && sb.charAt(0) == '0') {
            sb.deleteCharAt(0);
        }

        return sb.length() == 0 ? "0" : sb.toString();
    }
}

File: 402-remove-k-digits/remove-k-digits.cpp
class Solution {
public:
    string removeKdigits(string num, int k) {
        int n = num.size();
        stack<char> st;
        for(int i = 0;i < n;i++){
            while(!st.empty() && k>0 && (st.top()-'0')>(num[i]-'0')){
                st.pop();
                k--;
            }
            st.push(num[i]);
        }
        while(k>0){
            st.pop();
            k--;
        }
        if(st.empty()) return "0"; 
        string res = "";
        while(!st.empty()){
            //res = res + st.top(); 
            res.push_back(st.top());
            st.pop();
        }
        while(res.size() > 1 && res.back() == '0') {
            res.pop_back();
        }
        reverse(res.begin(),res.end());
        return res.empty() ? "0" : res;
    }
};